tests/test_pygtracker.py

```diff
@@ -18,9 +20,80 @@ def test_version():
 
 # Start tests for generate_course_statistics
 
+
+def test_generate_course_statistics_input_not_exist():
+    tracker = pygtracker.GradeTracker()
+    with raises(ValueError):
+        tracker.generate_course_statistics(course_ids=[500, "511"])
+
+
+def test_generate_course_statistics_input_type1():
+    tracker = pygtracker.GradeTracker()
+    with raises(TypeError):
+        tracker.generate_course_statistics(course_ids=[True, "511"])
+
+
+def test_generate_course_statistics_input_type2():
+    tracker = pygtracker.GradeTracker()
+    with raises(TypeError):
+        tracker.generate_course_statistics(course_ids="511")
+
+
+# the output is equal to the statistics summary calculate mannually
+# def test_generate_course_statistics_equal():
+# 511_mean = num1
+# 522_median = num2
+
+# tracker = pygtracker.GradeTracker()
+# output = tracker.generate_course_statistics()
+
+# assert output[output["course_id"]=="511"][1] == 511_mean
+# assert output[output["course_id"]=="522"][3] == 522_median
+
+# check the columns of the dataframe that come of the function
+
+
+def test_generate_course_statistics_output_match():
+    tracker = pygtracker.GradeTracker()
+    row_num = len(tracker.generate_course_statistics(course_ids=["511", "522"]))
+    columns_list = tracker.generate_course_statistics(course_ids=["511", "522"]).columns
+    assert columns_list[0] == "course_id"
+    assert columns_list[1] == "mean"
+    assert columns_list[2] == "1st-quantile"
+    assert columns_list[3] == "median"
+    assert columns_list[4] == "3rd-quantile"
+    assert row_num == 2
+
+
 # End tests for generate_course_statistics
 
 # Start tests for rank_courses
+def test_rank_courses_input_method_type():
+    tracker = pygtracker.GradeTracker()
+    with raises(TypeError):
+        tracker.rank_courses(descending=1)
+
+
+def test_rank_courses_input_method():
+    tracker = pygtracker.GradeTracker()
+    with raises(ValueError):
+        tracker.rank_courses(method="avg")
+
+
+def test_rank_courses_order():
+    tracker = pygtracker.GradeTracker()
+    assert_series_equal(
+        tracker.rank_courses(descending=True).iloc[-1],
+        tracker.rank_courses(descending=False).iloc[0],
```

Comment:
You can't test like this... what if the real data has 100 courses? You can test the last one from the ascending order is same as the first one from the descending order

Reply:
I think I already did what you described here. Line 86 is the last one from the descending order and Line 87 is the first one from the ascending order. `assert_series_equal` function will test if they are equal.

Reply:
What I mean is that:
- Assume your data has 2 courses
- By manually calculating the average grade for each course, you know that 511 > 522
- So you create these 2 tests:

1. if sorted by ascending order, 522 will be in the first row, 511 in the second row. And their average grades will match what you calculate manually
2. if sorted by descending order, 511 will be in the first row, 522 will be in the second row....

If you test like this, what if you make a mistake in the main function that returns the below?
"511", 90
"511", 90

The test still passes, but it should be able to capture the failure here...

Reply:
[-1] means the last row in data frame. So it will always return the last row of descending order and compare with the first row of ascending order, even if there are 100 or more courses.

Reply:
Yes... but if your function returns the data frame as I mentioned above. Will your test still pass? But should it pass in the first place?

Reply:
One rule of testing is that you test the output of your function with something that you can verify, for example a fixed number. If the test passes, it means the output is as expected.

You should avoid testing something that is uncertain (the first row in ascending order) with something that is uncertain (the last row in descending order).

Reply:
If I made a mistake in main function, the test for value will be failed. That is not the purpose of this test functions.